Mathlib/Algebra/GroupPower/Lemmas.lean

```diff
@@ -2,35 +2,1316 @@
+#check Nat.cast_pow
+
+/- Porting note: removed `@[simp]` because of linter error:
+
+Left-hand side simplifies from
+  ↑(Int.natAbs x ^ 2)
+to
+  ↑(Int.natAbs x) ^ 2
+using
+  simp only [Nat.cast_pow]
+Try to change the left-hand side to the simplified term!
+
+- Porting note: because of `HMul` more type annotations are needed than in the Lean 3 version.
+-/
+-- @[simp]
+theorem nat_abs_sq (x : ℤ) : ((x.natAbs ^ 2 : ℕ) : ℤ) = x ^ 2 := by rw [sq, Int.natAbs_mul_self, sq]
+#align int.nat_abs_sq Int.nat_abs_sq
+
+alias nat_abs_sq ← nat_abs_pow_two
+
+theorem abs_le_self_sq (a : ℤ) : (Int.natAbs a : ℤ) ≤ a ^ 2 := by
+  rw [← Int.nat_abs_sq a, sq]
+  norm_cast
+  apply Nat.le_mul_self
+#align int.abs_le_self_sq Int.abs_le_self_sq
+
+alias abs_le_self_sq ← abs_le_self_pow_two
+
+theorem le_self_sq (b : ℤ) : b ≤ b ^ 2 :=
+  le_trans le_natAbs (abs_le_self_sq _)
+#align int.le_self_sq Int.le_self_sq
+
+alias le_self_sq ← le_self_pow_two
+
+theorem pow_right_injective {x : ℤ} (h : 1 < x.natAbs) : Function.Injective ((· ^ ·) x : ℕ → ℤ) :=
+  by
+  suffices Function.Injective (natAbs ∘ ((· ^ ·) x : ℕ → ℤ)) by
+    exact Function.Injective.of_comp this
+  convert Nat.pow_right_injective h
+  ext n
+  rw [Function.comp_apply, nat_abs_pow]
+#align int.pow_right_injective Int.pow_right_injective
+
+end Int
+
+variable (M G A)
+
+-- Porting note: dot notation was not working so a roundabout way was used.
+-- Also the components of the structure are given explicitly (maybe possible to avoid this).
+/-- Monoid homomorphisms from `multiplicative ℕ` are defined by the image
+of `multiplicative.of_add 1`. -/
+def powersHom [Monoid M] :
+    M ≃
+      (Multiplicative ℕ →* M) where
+  toFun : M → Multiplicative ℕ →* M :=
+  fun x =>
+    { toFun := fun n : ℕ  =>
+        let n' := Multiplicative.toAdd n
+        x ^ n'
+      map_one' := by
+        convert pow_zero x
+      map_mul' :=  fun m n => pow_add x m n
+        }
+  invFun f := f (Multiplicative.ofAdd 1)
+  left_inv := pow_one
+  right_inv f := MonoidHom.ext fun n => by simp [← f.map_pow, ← ofAdd_nsmul]
+#align powers_hom powersHom
+
+/-- Monoid homomorphisms from `multiplicative ℤ` are defined by the image
+of `multiplicative.of_add 1`. -/
+def zpowersHom [Group G] :
+    G ≃
+      (Multiplicative ℤ →*
+        G) where
+  toFun: G → Multiplicative ℤ →* G :=
+  fun x => {
+    toFun := fun n : ℤ  =>
+      let n' := Multiplicative.toAdd n
+      x ^ n'
+    map_one' := zpow_zero x
+    map_mul' := fun m n => zpow_add x m n
+    }
+  invFun f := f (Multiplicative.ofAdd 1)
+  left_inv := zpow_one
+  right_inv f := MonoidHom.ext fun n => by
+    simp [← f.map_zpow, ← ofAdd_zsmul]
+#align zpowers_hom zpowersHom
+
+/-- Additive homomorphisms from `ℕ` are defined by the image of `1`. -/
+def multiplesHom [AddMonoid A] :
+    A ≃
+      (ℕ →+
+        A) where
+  toFun : A → ℕ →+ A :=
+  fun x =>
+    {
+    toFun := fun n => n • x
+    map_zero' := zero_nsmul x
+    map_add' := fun _ _ => add_nsmul _ _ _
+  }
```

Comment:
I can't say I understand the spacing conventions in Lean 4 but I doubt the `{}` are supposed to be misaligned (don't ask me which one to change though...)

Reply:
changed, but don't know if this is correct